Add tinyCBOR external library and apply mynewt upstream patch from intel/tinycbor#83

vrahane opened the pull request:
- Add tinycbor as an external library commit: intel/tinycbor@497066e
- This is needed by MCUmgr and config subsystem in zephyr
- Pulling in code from intel/tinycbor#83 as a patch to zephyr's ext/tinycbor. This is to facilitate the 
  use of chained buffers functionality for tinycbor while it is in development on 
  https://github.com/intel/tinycbor
- Bring tinycbor up to date with mynewt tinycbor
- Changing implementation of cbor_encoder_get_extra_bytes_needed() and 
   cbor_encoder_get_buffer_size() as part of cbor_buf_writer APIs
- Move bytes_needed field from CborEncoder to the buf writer, this is needed by the tests mainly.
- Fix cbor_buf_cmp to do memcmp and return complemented result iterate_string_chunks(): fixing 
  NULL compare at the end of string and moving it out of the iterate_string_chunks(). This is to 
  avoid buffer specific parser calls in the function
- cbor_value_get_next_byte() is removed in mynewt version of tinycbor, so, we track offsets of the 
  buffer which can be used for comparison in the parser tests instead of calculating the offset
- Making the decoder and parser APIs backwards compatible
- Adding encoder writer and parser reader as part of the encoder and parser structure. This is to 
  make the encoder and parser use new function of encoder_writer and decoder_reader without 
  breaking backwards compatibility.
- Making the old API use flat buffers by default
- Adding APIs for initializing encoder and parser with custom writer and reader
- Make the default writer and reader conditional based on NO_DFLT_READER/WRITER define. This 
  is because we want a default reader/writer to avoid API changes.
- Use it->offset instead of it->ptr to track buffer offsets
- Update resolve_indicator() static api parameters to use cbor value and access offsets instead of 
  taking pointers as input parameters
- In validate_container() do a byte by byte comparison instead of memcmp since we no longer 
   have access to the buffer directly
- Also, use offsets instead of pointers to validate sorted maps
- Added a new define for conditionally compiling in float support (NO_FLOAT_SUPPORT).
- This is because we want the float support to be compiled in by default.
- Use static_assert macro instead of Static_assert. Changed to avoid build failures.
- Add api to get string chunk, this is a callback which can be used by buffer implementations to 
  grab a string that is divided in chunks which spans across multiple chained buffers

Fixes #3081

Signed-off-by: Vipul V Rahane <vipul@runtime.io>

carlescufi commented:
@vrahane I think we should also include integration with Kconfig and CMake in this PR. Otherwise this will not be usable by Zephyr modules. We cannot build with make because that does not work on Windows. Something like this: https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/CMakeLists.txt#L7

carlescufi commented:
@vrahane in the first commit message, can you remove `- This is needed by MCUmgr and config subsystem in zephyr` and replace it with a very brief description of what tinyCBOR is/does?

carlescufi commented:
I understand having the full repo makes it easier later to rebase, but I'm not sure we should include things like `.gitignore` and similar.

carlescufi reviewed (changes requested):
This should live in `ext/lib/tinycbor` and not in `ext/tinycbor`

vanwinkeljan commented:
Should mynewt changes not first be upstreamed to intel/tinycbor before taking it in to zephyr repo?
Now you would be taking in a fork instead of the upstream project.

carlescufi commented:
@vanwinkeljan they should, but the won't make it in time for Zephyr 1.11. This is a stopgap solution until @thiagomacieira merges those.

thiagomacieira commented:
I'll join zephyr-devel and ask questions. I want to have native support for `net_buf` and `net_pkt` in TinyCBOR, but I need to understand a few things first.

carlescufi commented:
@thiagomacieira thanks, although the work so far has been done by Mynewt developers. I'll ask them to join zephyr-devel as well.

vrahane commented:
@carlescufi I have made changes as per the review and the builds are now successful as well. Thank you.

mbolivar commented:
@vrahane Cool!

One thing, though. You need to use interactive rebase to edit the pull request history into a coherent set of patches, rather than addressing review with commits at the end of the series, like the "Make changes as per PR review" one that is present right now.

Here is the workflow document: http://docs.zephyrproject.org/contribute/contribute_guidelines.html#contribution-workflow

vrahane commented:
@mbolivar I generally prefer rebasing as well. The only issue here is that the first commit is the vanilla version of tinycbor, second applies the mynewt patch on top of it and the third one specifically adds Kconfig, changes directories and deletes git metadata files from the library. If the patch(second commit) were to be ever reverted, minimal changes would have to be made to the CMake files, Kconfig and menuconfig which is added by the third commit. Hence, the third commit. Maybe I should name it name it differently instead of just a PR review changes, Any thoughts appreciated ?

carlescufi commented:
@vrahane That is fine but you cannot name a commit "Make changes as per PR review". You need to name commits after what they contain.

carlescufi commented:
@vrahane also the removal of non-needed files should not really be in this 3rd commit. Those files should really never have been included in the first place, we can always trace back  the diffs from the imported files to the next commit changes with the patch from the TinyCBOR upstream PR regardless of these files being included or not. In fact, we only need the diffs on the relevant files for Zephyr.

nashif reviewed (changes requested):
Can you please just import the files we need to build the library and drop everything else? We should not include any of the CI, IDE, project scripts and such, those will just trigger all kind of violations when we do license scanning and IP management. Just keep the C and header files we need for building and drop everything else. Maybe later when we have better repository management we can import complete tree, but not now.

vrahane commented:
Thank you @nashif and @carlescufi for going through it. As per your reviews and zephyr policies, I have done the following:
- Added KConfig and CMakeList files for configuration and build
- Added tinycbor to ext/lib/encoding
- Deleted .gitignore and .gitattributes
- Removed tools, tests, examples, CI scripts, Makefiles as they are not really needed for building the library with Zephyr
- Rebased and squashed unnecessary commits made for cleaning up the library

carlescufi commented:
@vrahane I think it would be cleaner if we did this in a slightly different way:

Commit 1: Add the required TinyCBOR files **only**, without the 83 patch
Commit 2: Patch according to 83, again only the required files
Commit 3: Add Kconfig and CMakeLists 

That said I am OK with merging this as-is if @nashif is as well

carlescufi commented:
@vrahane also can you please remove everything that is not required for building? Files like `Doxyfile` and `TODO`, etc.

vrahane committed:
ext: lib: tinycbor: Apply req patch to 0.5.0-beta1

For a full description of changes see the commit message
URL: https://github.com/zephyrproject-rtos/tinycbor/tree/zephyr

Signed-off-by: Vipul Rahane <vipulrahane@apache.org>

vrahane commented:
@carlescufi and @nashif: Made changes as per our discussion.

nashif commented on ext/lib/encoding/tinycbor/Kconfig:
```diff
@@ -0,0 +1,89 @@
+
+config CBOR_NO_DFLT_WRITER
+	bool
+  depends on TINYCBOR
```
all of those "depends on" are misaligned. You can also add if TINYCBOR just before this and close with endif at the bottom and get rid of them all.

vrahane replied:
Sorry for missing that. Will fix it.

nashif commented on ext/lib/encoding/tinycbor/src/compilersupport_p.h:
```diff
@@ -0,0 +1,263 @@
+#  include <assert.h>
+#endif
+#include <float.h>
+#include <math.h>
```
zephyr does not have this in its minimal libc, so you will have to make tinycbor depend on newlib libc, otherwise build will fail. If we had a simple test using the library, this would have been identified as an issue earlier.

carlescufi replied:
Agreed, mcumgr also depends on newlib libc. CC @nvlsianpu @ccollins476ad

thiagomacieira replied:
float.h should be available everywhere. It's required by the C standard even in freestanding implementations and it is provided by gcc, not the libc.

math.h is used in the `decode_half` function and that one only uses `ldexp`. I will accept a Pull Request that reimplements the function without `ldexp`, like `encode_half` above. If anyone wants to contribute an ARM `__fp16` implementation like the x86 F16C, I'll also accept it.

thiagomacieira replied:
assert.h also seems to be in the toolchain:

```
$ find /opt/zephyr-sdk/ -name assert.h
/opt/zephyr-sdk/sysroots/mips32r2-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/arc-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/riscv32-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/i586-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/nios2-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/iamcu-zephyr-elfiamcu/usr/include/assert.h
/opt/zephyr-sdk/sysroots/xtensa-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/armv5-zephyr-eabi/usr/include/assert.h
```

It is compliant with C11. It does call a function called `__assert_func` on failed assertions, but if you don't want hat you can always compile with NDEBUG.

carlescufi replied:
@ccollins476ad @vrahane you mentioned you are building this without newlib?

ccollins476ad replied:
@carlescufi mcumgr doesn't require any of the TinyCBOR C files that include `math.h` (`cbortojson.c`, `cborpretty.c`, and `cborvalidation.c`)[*].  Removing these files from the TinyCBOR `CMakeLists.txt` file allowed a successful build without pulling in newlib.

Since TinyCBOR is being added as an external zephyr library, and not just an external mcumgr library, we can't just remove these files from the `CMakeLists.txt` file.  Perhaps a new Kconfig setting to exclude these files is required.

[*] `compilersupport_p.h` also includes `math.h`.  I'm not sure why this didn't cause build failures, as several source files include this header.  I don't have a suitable machine available at the moment, but I will investigate this later today and include my findings here.

vrahane replied:
I agree with @ccollins476ad above. We might have to investigate this further but adding a config option to depend on does seem like the right approach.

ccollins476ad replied:
I looked at this again, and it seems my compiler[1] has no problem with `#include <math.h>`, even when `CONFIG_NEWLIB_LIBC` is disabled.

I am pretty sure I used to get a build error due to the missing `math.h` header, but now I can't reproduce it.  Maybe I just remembered incorrectly...

[1] `gcc version 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496] (GNU Tools for ARM Embedded Processors)`

carlescufi replied:
@ccollins476ad can you please try with the Zephyr SDK: http://docs.zephyrproject.org/getting_started/installation_linux.html#installing-the-zephyr-software-development-kit

ccollins476ad replied:
Unfortunately, I do get build errors with the Zephyr SDK:
```
In file included from /home/ccollins/repos/zephyr/ext/lib/encoding/tinycbor/src/cbor_buf_reader.c:26:0:
/home/ccollins/repos/zephyr/ext/lib/encoding/tinycbor/src/compilersupport_p.h:44:18: fatal error: math.h: No such file or directory                                    
 #include <math.h>                                                                                                                                                     
                  ^                                                                                                                                                    
compilation terminated.                                                                                                                                                
```

thiagomacieira replied:
I don't see math.h anywhere in the Zephyr sources or SDK. Like I said, the only function used is ldexp and that was only so I wouldn't depend on IEEE 754 internals (like the other function does).

In any case, as  @ccollins476ad said, the `decode_half` function is only used in the cborpretty.c, cborvalidation.c, and cbortojson.c, which aren't used, so you can just erase the function and the include, for now.

(That would make it my problem to fix when integrating with Zephyr...)

carlescufi commented:
@vrahane Also please change the first commit so it only contains the `src/` folder, and not the whole of TinyCBOR.

nashif commented:
@thiagomacieira  do you have c based tests for tinycbor that we could adapt for Zephyr? All I can see in the tree are QT based tests

thiagomacieira commented:
Not yet, no. That's why I asked you last Friday for how to use Travis with Zephyr. I need to write a couple, though I will focus only on the Zephyr-specific code (yet to be written) that integrates with `net_buf`. I won't re-do the entire suite, though I realise the current runs don't test any 32-bit build.

vrahane committed:
ext: lib: tinycbor: Add tinyCBOR as an ext lib

The TinyCBOR library is a small Concise Binary Object
Representation (CBOR) encoder and decoder library, optimized for
very fast operation with very small footprint.

Origin: TinyCBOR
License: MIT
URL: https://github.com/intel/tinycbor
Version: 0.5.0-beta1
commit: 497066ee87dd54341adaa1195bf15ad11ee33b20
Purpose: Introduction of TinyCBOR
Maintained-by: External

Signed-off-by: Vipul Rahane <vipulrahane@apache.org>

thiagomacieira commented on ext/lib/encoding/tinycbor/Kconfig:
```diff
@@ -0,0 +1,85 @@
+
+config CBOR_ENCODER_NO_CHECK_USER
+	bool
+	prompt "No check user for cbor encoder"
```
Better text: This option controls whether the CborEncoder checks whether the user passed valid arguments.

vrahane replied:
It is a prompt message, hence you cannot put a long descriptive text there.

thiagomacieira commented on ext/lib/encoding/tinycbor/Kconfig:
```diff
@@ -0,0 +1,85 @@
+
+if TINYCBOR
+
+config CBOR_NO_DFLT_WRITER
```
Maybe find a better name for this? At least get rid of the acronym.

I've been calling this the ability to encode to a linear buffer, as opposed to a `net_buf`.

vrahane replied:
Any suggestion ?

thiagomacieira commented on ext/lib/encoding/tinycbor/Kconfig:
```diff
@@ -0,0 +1,85 @@
+	help
+	  This option enables floating point support.
+
+config CBOR_NO_HALF_FLOAT_TYPE
```
This option should control the `decode_half` function we've been having problems with too.

ritikakapade replied:
Ok, this fixes the half float issue and does not pull in newlib libc.

thiagomacieira commented on ext/lib/encoding/tinycbor/Kconfig:
```diff
@@ -0,0 +1,85 @@
+	help
+	  This option enables half float type support.
+
+config CBOR_WITHOUT_OPEN_MEMSTREAM
```
Unnecessary, since there is no open_memstream support in Zephyr. It's a Glibc extension to POSIX, though I think Apple recently added to their libc too.

vrahane replied:
Are you suggesting we should remove it ?

thiagomacieira commented on ext/lib/encoding/tinycbor/Kconfig:
```diff
@@ -0,0 +1,85 @@
+config CBOR_WITHOUT_OPEN_MEMSTREAM
+	bool
+	prompt "without open memstream"
+	default n
```
Are you sure? Note that the #define is negative ("without'), so this makes a double negative and you're saying **with** `open_memstream`, which Zephyr doesn't have.

vrahane replied:
Yes, this needs to get fixed. Thank you.

vrahane commented:
I discussed this with @ccollins476ad and we tested a few things out on the Zephyr-SDK. Based on the findings, I have made two commits above:

2635b19: ext: lib: tinycbor: make HF type conditional 
This commit does the following:
- Make half float encode/decode conditional based on `CBOR_NO_HALF_FLOAT_TYPE`
- Changes defaults for `CBOR_WITHOUT_OPEN_MEMSTREAM` and
  `CBOR_NO_HALF_FLOAT_TYPE` to `y` so that half float support and memstream support does 
  not get compiled in as it is not needed for normal usage of the library

2edd586: ext: lib: tinycbor: Fix build issues
- This fixes build issues that we have been discussing on this PR wrt newlib libc and float support
- src/cborpretty.c, src/cbortojson.c and src/cborvalidation.c
  conditionally include math.h and half float type support. 
  They are not specified in the CMakeList files because of the following reasons: 
  i. The following source files requires `cbor_value_to_pretty_advance()`, which is hosted-only.
 - `src/cborpretty_stdio.c`
 - `src/cbortojson.c `         

 ii. The following source file  requires glibc which is not part of Zephyr and hence it is compiled 
     out by default:
 - `src/open_memstream.c` 
  
- Change default settings for floating type support and half
  float type support to be disabled by default to avoid newlib libc from
  getting compiled in by default `CBOR_NO_HALF_FLOAT_TYPE` and 
 `CBOR_NO_FLOATING_TYPE` is set to `y`.
- Conditionally include `math.h` in `src/compilersupport_p.h` to avoid
  newlib libc from getting compiled in using `CBOR_NO_FLOATING_TYPE`
- Conditionally compile src/cborparser_dup_string.c if newlib libc is
  compiled in using `CBOR_NO_FLOATING_TYPE`

This solves most of the issues if not all. I have also changed prompt messages to make it more aesthetic. Please let me know if this works. I have kept it as separate commits so that everybody can take a look at it. I can further combine it into a single commit or maybe split it into multiple based on what feedback I get.

thiagomacieira commented:
The changes look good, with minor changes. I'd like the half and full-FP support changes in TinyCBOR.

carlescufi commented:
@vrahane Thanks for all the changes related to building with and without newlib. These commits that modify `src/` need to go go to the official Zephyr upstream fork like you did for the commit with all the changes: https://github.com/zephyrproject-rtos/tinycbor/tree/zephyr, so we can keep all changes to TinyCBOR itself there. Finally I would prefer if you could squash the last 2 commits together, we really don't want a commit that says "Fix the build".

vrahane commented:
Sorry @SebastianBoe, did not mean to add you as a reviewer.

vrahane commented:
@carlescufi I have made the above half/full FP changes to the PR to https://github.com/zephyrproject-rtos/tinycbor/tree/zephyr after squashing.

vrahane committed:
ext: lib: tinycbor: half/full-FP support changes

- Make half float encode/decode conditional
- src/cborpretty.c, src/cbortojson.c and src/cborvalidation.c
  conditionally include math.h and half float type support
- Conditionally include math.h in src/compilersupport_p.h to avoid
  newlib libc from getting compiled in
- Conditionally compile src/cborparser_dup_string.c if newlib libc is
  compiled in

Signed-off-by: Vipul Rahane <vipulrahane@apache.org>

vrahane committed:
ext: lib: tinycbor: Add req Zephyr specific files

Include both build an configuration files required to build
TinyCBOR with Zephyr.

Signed-off-by: Vipul Rahane <vipulrahane@apache.org>

carlescufi reviewed (approved):
Looks good to me now

carlescufi commented:
@nashif mind taking another look at this?
